app/src/lib/git/stash.ts

```diff
@@ -113,10 +111,21 @@ export function createDesktopStashMessage(branchName: string) {
  */
 export async function createDesktopStashEntry(
   repository: Repository,
-  branchName: string
+  branchName: string,
+  untrackedFilesToStage: ReadonlyArray<string>
 ): Promise<true> {
+  // We must ensure that no untracked files are present before stashing
+  // See https://github.com/desktop/desktop/pull/8085
+  if (untrackedFilesToStage.length) {
+    await git(
+      ['add', ...untrackedFilesToStage],
+      repository.path,
+      'stageUntrackedFilesBeforeStashing'
+    )
```

Comment:
A few notes: 

I'm intentionally choosing to do the staging of untracked files in this method (as opposed to in AppStore prior to calling this) because it's a necessary step of stashing correctly. I just deleted a bunch of code in this file that was handling the case of untracked files in the stash, so I want it to be nearly impossible to accidentally stash without first staging all untracked files. 

After discussing with @outofambit, I decided to pass untracked files in rather than shell out to Git and get this information. The reason is that stashing already takes a long time and requires several git calls, and we want to don't want to make it even slower with another git call since we can avoid it. 

I considered staging all files with a `git add .` for the sake of both simplicity and consistency. If we're gonna muck with the user's index by staging something without consulting them, it's probably better to do so in an all-or-nothing way where we stage all of their files so there's consistency and predictability wrt what's staged. But, once again, I decided to opt for the more performant approach and stage only the untracked files. Command line profiling shows that staging individual files is faster than `git add .`

And I decided not to use the `stageFiles` method in `update-index.ts` because it does a lot of unnecessary stuff, including multiple `git update-index` calls.

Reply:
documented in #8345

Reply:
Turns out that using `stageFiles` is less heavy-handed than I originally thought. We call the `updateIndex` method a few times, but all but one will be a no-op because we're passing empty file arrays. So 👍 to using `stageFiles`. 

And thanks for providing the valuable context @niik! Glad you caught this and saved us future bites by the Window's command line argument length cap 😌 

Addressed in https://github.com/desktop/desktop/pull/8363